src/celeritas/global/detail/AlongStepActionImpl.hh

```diff
@@ -234,6 +235,9 @@ inline CELER_FUNCTION void along_step_track(CoreTrackView const& track)
 
     // Override step limit with whatever action/step changes we applied here
     sim.force_step_limit(step_limit);
+
+    // Increment the step counter
+    sim.increment_num_steps();
```

Comment:
I am not sure whether this is the right place to increase the step counter since any of along step actions (either
range limiter or msc) may not be selected for the process governing the current step - well the range limiter is a part of pre-step though in our implementation.  So, should it be increased either when msc is selected, or inside somewhere discrete_select_track if msc (and the range limiter?) is not select.

Reply:
The end of "along step" seems like the right place to step counter, right? Any track that is active needs to have its step counter incremented, and (aside from the "zero step" special case where we now increment it) this will be called for every active track.

Reply:
To me, places where we should increase the step counter (exclusively) are 1) when msc is selected, i.e, [inside the block](https://github.com/sethrj/celeritas/blob/safety-msc/src/celeritas/global/detail/AlongStepActionImpl.hh#L115) and 2) when the geometry limited step is selected, i.e, [inside the block](https://github.com/sethrj/celeritas/blob/safety-msc/src/celeritas/global/detail/AlongStepActionImpl.hh#L129), and 3) [the inside discrete_select_track](https://github.com/sethrj/celeritas/blob/safety-msc/src/celeritas/phys/detail/DiscreteSelectActionImpl.hh#L45).  Of course, it also need to check whether either msc or geometry limited step is selected at [the line](https://github.com/sethrj/celeritas/blob/safety-msc/src/celeritas/phys/detail/DiscreteSelectActionImpl.hh#L35).  So, not sure whether putting the `increment_num_steps()` at the end of along step is equivalent to the logic above, i.e, make sure whether the step counter is neither over counted nor under counted, but increased once and only once per alived step.

Reply:
I think the end of AlongStep is going to be the cleanest place to put it: it's the only place (barring the one "return if zero step length") that is going to be guaranteed to be called before the final particle processing... having to put it in multiple different kernels risks forgetting to put it somewhere. Adding it to more locations (such as when the step limiter is selected, which at that point can still be overridden by slowing down) increases the risk of errors.

I guess it really matters most depending on what actually uses the step counter: and since the only things right now are the diagnostics and the MSC "first step" exception, and nothing in physics (or boundary crossing) does. So let's leave it here as the last thing that happens in "along step", and we can make sure it's documented and can later change it as needed.